[ROCm] Support for Whisper v1 with Aiter Unified Attention and Aiter Flash Attention

apinge opened the pull request:
## Purpose
This PR enables broader attention backend support for Whisper v1 on ROCm platform.
Building on the existing Triton backend PR #28346 , it introduces:

Aiter Unified Attention
Aiter Flash Attention

This change depends on modifications from the Triton backend PR. Since both PRs modify the same file (`vllm/v1/worker/utils.py`).

## Test Plan
Whisper v1 on ROCm with Aiter backends requires the latest Aiter version, tested with commit 7639e55

```bash
export CUDA_VISIBLE_DEVICES=0,1
export VLLM_USE_ROCM_AITER=1
# test aiter unified attention
export VLLM_ATTENTION_BACKEND=ROCM_AITER_UNIFIED_ATTN
 pytest ./tests/models/multimodal/generation/test_whisper.py 
 # test aiter flash attention
export VLLM_ATTENTION_BACKEND=ROCM_AITER_FA
 pytest ./tests/models/multimodal/generation/test_whisper.py
```
## Test Result

### Result of Aiter Unified Attention
```
============================================================================================ test session starts =============================================================================================
platform linux -- Python 3.12.11, pytest-8.4.2, pluggy-1.6.0
rootdir: /root/workspace/vllm_master_251109
configfile: pyproject.toml
plugins: asyncio-1.2.0, anyio-4.11.0
asyncio: mode=Mode.STRICT, debug=False, asyncio_default_fixture_loop_scope=None, asyncio_default_test_loop_scope=function
collected 3 items

models/multimodal/generation/test_whisper.py ...                                                                                                                                                       [100%]

============================================================================================== warnings summary ==============================================================================================
<frozen importlib._bootstrap>:488
  <frozen importlib._bootstrap>:488: DeprecationWarning: builtin type SwigPyPacked has no __module__ attribute

<frozen importlib._bootstrap>:488
  <frozen importlib._bootstrap>:488: DeprecationWarning: builtin type SwigPyObject has no __module__ attribute

../../../../usr/local/lib/python3.12/dist-packages/audioread/rawread.py:16
  /usr/local/lib/python3.12/dist-packages/audioread/rawread.py:16: DeprecationWarning: 'aifc' is deprecated and slated for removal in Python 3.13
    import aifc

../../../../usr/local/lib/python3.12/dist-packages/audioread/rawread.py:17
  /usr/local/lib/python3.12/dist-packages/audioread/rawread.py:17: DeprecationWarning: 'audioop' is deprecated and slated for removal in Python 3.13
    import audioop

../../../../usr/local/lib/python3.12/dist-packages/audioread/rawread.py:19
  /usr/local/lib/python3.12/dist-packages/audioread/rawread.py:19: DeprecationWarning: 'sunau' is deprecated and slated for removal in Python 3.13
    import sunau

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
======================================================================================= 3 passed, 5 warnings in 31.29s =======================================================================================
sys:1: DeprecationWarning: builtin type swigvarlink has no __module__ attribute
```
### Result for Aiter Flash Attention
```
pytest ./models/multimodal/generation/test_whisper.py
============================================================================================ test session starts =============================================================================================
platform linux -- Python 3.12.11, pytest-8.4.2, pluggy-1.6.0
rootdir: /root/workspace/vllm_master_251109
configfile: pyproject.toml
plugins: asyncio-1.2.0, anyio-4.11.0
asyncio: mode=Mode.STRICT, debug=False, asyncio_default_fixture_loop_scope=None, asyncio_default_test_loop_scope=function
collected 3 items

models/multimodal/generation/test_whisper.py ...                                                                                                                                                       [100%]

============================================================================================== warnings summary ==============================================================================================
<frozen importlib._bootstrap>:488
  <frozen importlib._bootstrap>:488: DeprecationWarning: builtin type SwigPyPacked has no __module__ attribute

<frozen importlib._bootstrap>:488
  <frozen importlib._bootstrap>:488: DeprecationWarning: builtin type SwigPyObject has no __module__ attribute

../../../../usr/local/lib/python3.12/dist-packages/audioread/rawread.py:16
  /usr/local/lib/python3.12/dist-packages/audioread/rawread.py:16: DeprecationWarning: 'aifc' is deprecated and slated for removal in Python 3.13
    import aifc

../../../../usr/local/lib/python3.12/dist-packages/audioread/rawread.py:17
  /usr/local/lib/python3.12/dist-packages/audioread/rawread.py:17: DeprecationWarning: 'audioop' is deprecated and slated for removal in Python 3.13
    import audioop

../../../../usr/local/lib/python3.12/dist-packages/audioread/rawread.py:19
  /usr/local/lib/python3.12/dist-packages/audioread/rawread.py:19: DeprecationWarning: 'sunau' is deprecated and slated for removal in Python 3.13
    import sunau

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
======================================================================================= 3 passed, 5 warnings in 29.38s =======================================================================================
sys:1: DeprecationWarning: builtin type swigvarlink has no __module__ attribute

```
---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

apinge committed:
add support for whisper v1 using aiter unified attention and aiter flash attention

Signed-off-by: apinge <Tong.Qiu2@amd.com>

apinge committed:
update key and value for the None condition

Signed-off-by: apinge <Tong.Qiu2@amd.com>

apinge committed:
update format for rocm_aiter_unified_attn.py

Signed-off-by: apinge <Tong.Qiu2@amd.com>

apinge committed:
update format

Signed-off-by: apinge <Tong.Qiu2@amd.com>

apinge commented:
> @apinge I understand that you have stated that we need to use latest AITER commit. Is there any chance that this work with the AITER version in
> 
> https://github.com/vllm-project/vllm/blob/6b2b9fd934af2bf59f749aa04ac739fd9c7d29f7/docker/Dockerfile.rocm_base#L10
> 
> ?

This aiter version works fine, provided that #28383  is applied.

tjtanaa commented:
There are the same review questions related to PR https://github.com/vllm-project/vllm/pull/28346 . We will wait for the other PR issue to sort out the issues.

apinge commented:
> There are the same review questions related to PR #28346 . We will wait for the other PR issue to sort out the issues.

I found that the changes addressing the review questions are causing an accuracy problem — I’ve left a comment in #28346 .

Also, testing shows that the Aiter Flash Attention backend can hit a NaN issue for some prompts, and this issue has been fixed in PR #28670 .

micah-wil commented:
Hi @apinge, could you check the CI failure? It kind of just looks like the failing test is getting rate limited by Wikimedia servers or something, but it has failed a few times now.

cc @tjtanaa

DarkLight1337 committed:
Merge branch 'main' into whisper_v1_aiter

apinge committed:
Merge branch 'main' into whisper_v1_aiter
